src/crypto/bip324_suite.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2019-2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <crypto/bip324_suite.h>
+
+#include <crypto/common.h>
+#include <crypto/poly1305.h>
+#include <crypto/sha256.h>
+#include <support/cleanse.h>
+
+#include <assert.h>
+#include <cstring>
+#include <string.h>
+
+BIP324CipherSuite::~BIP324CipherSuite()
+{
+    memory_cleanse(key_P.data(), key_P.size());
+}
+
+void BIP324CipherSuite::Rekey()
+{
+    ChaCha20 rekey_c20(UCharCast(key_P.data()));
+    std::array<std::byte, NONCE_LENGTH> rekey_nonce;
+    memset(rekey_nonce.data(), 0xFF, 4);
+    memcpy(rekey_nonce.data() + 4, nonce.data() + 4, NONCE_LENGTH - 4);
+    rekey_c20.SetRFC8439Nonce(rekey_nonce);
+    rekey_c20.SeekRFC8439(1);
+    rekey_c20.Keystream(reinterpret_cast<unsigned char*>(key_P.data()), BIP324_KEY_LEN);
+}
+
+bool BIP324CipherSuite::Crypt(const Span<const std::byte> aad,
+                              const Span<const std::byte> input,
+                              Span<std::byte> output,
+                              BIP324HeaderFlags& flags, bool encrypt)
+{
+    // check buffer boundaries
+    if (
+        // if we encrypt, make sure the destination has the space for the encrypted length field, header, contents and MAC
+        (encrypt && (output.size() < BIP324_LENGTH_FIELD_LEN + BIP324_HEADER_LEN + input.size() + RFC8439_EXPANSION)) ||
+        // if we decrypt, make sure the source contains at least the encrypted header + mac and the destination has the space for the input - MAC - header
+        (!encrypt && (input.size() < BIP324_HEADER_LEN + RFC8439_EXPANSION || output.size() < input.size() - BIP324_HEADER_LEN - RFC8439_EXPANSION))) {
+        return false;
+    }
+
+    if (encrypt) {
+        // input is just the contents
+        // output will be encrypted contents length + encrypted (header and contents) + mac tag
+        uint32_t contents_len = input.size();
+        WriteLE32(reinterpret_cast<unsigned char*>(&contents_len), contents_len);
+
+        std::vector<std::byte> header_and_contents(BIP324_HEADER_LEN + input.size());
+
+        memcpy(header_and_contents.data(), &flags, BIP324_HEADER_LEN);
```

Comment:
```suggestion
        std::memcpy(header_and_contents.data(), &flags, BIP324_HEADER_LEN);
```
Same for memset etc.